File: typescript/src/canonical.ts
// Canonical JSON (RFC 8785 JCS approximation) — matches Python side for
// the value set used by our target APIs. See implementation-notes §1.1.

export function canonicalJSON(value: unknown): string {
  return stringify(value);
}

function stringify(v: unknown): string {
  if (v === null) return "null";
  if (typeof v === "boolean") return v ? "true" : "false";
  if (typeof v === "number") {
    if (!Number.isFinite(v)) {
      throw new Error("canonicalJSON: non-finite number is not allowed");
    }
    // JCS §3.2.2: use ECMAScript Number.prototype.toString.
    return String(v);
  }
  if (typeof v === "string") return JSON.stringify(v);
  if (Array.isArray(v)) {
    return "[" + v.map((x) => stringify(x)).join(",") + "]";
  }
  if (typeof v === "object") {
    const obj = v as Record<string, unknown>;
    const keys = Object.keys(obj).sort();
    return (
      "{" +
      keys
        .map((k) => JSON.stringify(k) + ":" + stringify(obj[k]))
        .join(",") +
      "}"
    );
  }
  throw new Error(`canonicalJSON: unsupported type ${typeof v}`);
}

File: python/src/agentreplay/canonical.py
"""Canonical JSON serialization (RFC 8785 JCS approximation).

Notes
-----
- CPython's ``json.dumps`` with ``sort_keys=True`` and ``separators=(",", ":")``
  produces a near-JCS form that matches TS ``JSON.stringify`` on canonical
  key-sorted objects for all values used by the v0.1 target APIs.
- Non-finite floats are rejected (`allow_nan=False`) because JCS forbids them.
- Unicode is emitted verbatim (``ensure_ascii=False``) to match JCS §3.2.1.
"""

from __future__ import annotations

import json
from typing import Any


def canonical_json(obj: Any) -> str:
    """Return an RFC 8785 (JCS)-style canonical JSON string for ``obj``.

    The serialization is stable across processes / OSes / Python versions.
    """
    return json.dumps(
        obj,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
        allow_nan=False,
    )
